### ISSUE TITLE
PHP : Segmentation fault

### ISSUE THREAD

**Issue by amereii** [google/protobuf#4757]
I have a **Segmentation fault** in PHP 7.1.17 , Centos 7 , libprotoc 3.5.1

## Step to reproduce

```
git clone https://github.com/RooyeKhat-Media/iGap-PHP-Client.git
cd iGap-PHP-Client/src/examples/
php getRoomList.php
```

> You can comment https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/d806654165da43b05db5bb867079f51a8716b03f/src/examples/getRoomList.php#L39-L42

In https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/d806654165da43b05db5bb867079f51a8716b03f/src/examples/getRoomList.php#L44 we have a **Segmentation fault** when new instance of `\Proto\UserLogin` is created

**Comment by amereii** [comment]
.proto file : http://irangit.com/core/iGap-API/blob/master/app/assets/proto/UserLogin.proto#L10-26

**Comment by amereii** [comment]
.proto file : https://github.com/RooyeKhat-Media/iGap-API/blob/master/app/assets/proto/UserLogin.proto#L10-L26

**Comment by amereii** [comment]
@TeBoring Do you have plan to fix it?

**Comment by TeBoring** [comment]
Sorry, I haven't started to look at it yet. Will investigate later this week

**Comment by lazeevv** [comment]
I have the same problem with php7.2.2, MacOS, libprotoc 3.5.1.
To reproduce:
```
<?php
require __DIR__ . '/../vendor/autoload.php';
$test = new class extends \Google\Protobuf\Internal\Message {};
```

**Comment by TeBoring** [comment]
@lazeevv you are not supposed to extend ```Message``` by yourself.

**Comment by TeBoring** [comment]
@amereii, I am not familiar with iGap-PHP-Client. With the info you give, I can hardly tell it's protobuf's problem. Could you remove all code that's not related to protobuf?

**Comment by amereii** [comment]
@TeBoring it is not related to **iGap-PHP-Client**
Just complite this proto file https://github.com/RooyeKhat-Media/iGap-API/blob/master/app/assets/proto/UserLogin.proto#L10-L26 and run the following code

```php
new \Proto\UserLogin();
```

**Comment by TeBoring** [comment]
Which files do you include?

**Comment by amereii** [comment]
It is very simple , Here is a step by step guide

# Option 1

* compile https://github.com/RooyeKhat-Media/iGap-API/blob/master/app/assets/proto/UserLogin.proto#L10-L26
* In your `composer.json` add and dump autoloader
```json
    "autoload": {
        "psr-4": {
            "Proto\\": "./Proto"
        }
    },
```
* create `repro.php`
```php
<?php
require __DIR__ . '/vendor/autoload.php';
new \Proto\UserLogin();
```
* run `php repro.php`

# Option 2 
* ```composer require rooyekhat/igap-php-client```
* create `index.php`
```php
require __DIR__ . '/vendor/autoload.php';
new \Proto\UserLogin();
```
* run `php index.php`

**Comment by TeBoring** [comment]
UserLogin.proto also import other proto files. Have you also compiled them?

**Comment by TeBoring** [comment]
Besides, where do u put generated GPBMetada files?

**Comment by amereii** [comment]
>UserLogin.proto also import other proto files. Have you also compiled them?
Yes , https://github.com/RooyeKhat-Media/iGap-PHP-Client/tree/master/src/Proto

>Besides, where do u put generated GPBMetada files?
https://github.com/RooyeKhat-Media/iGap-PHP-Client/tree/master/src/GPBMetadata

**Comment by TeBoring** [comment]
In your composer.json, have you configured psr-4 path for GPBMetadata?

**Comment by lazeevv** [comment]
@TeBoring yes, I know. It was just for reproducing the error.
I resolved the problem after your last comment, I just forgot to configure psr-4 path for GPBMetadata, thank you.
@amereii Are you sure that you ran `composer dumpautoload`?

**Comment by amereii** [comment]
@lazeevv 
> @amereii Are you sure that you ran composer dumpautoload?

Yes

@TeBoring I send you step by step guide to repro issue  , Why you don't follow it?

**Comment by TeBoring** [comment]
@amereii your reproduce step is coupled with other library, which is hard for me to tell what's the exact problem. Could you refine your reproducing step to use protobuf only?

**Comment by TeBoring** [comment]
@amereii in option1, step 1, what exactly you do to compile the proto file? In which working directory do you call protoc and what's the command to compile?
Ideally, I need a copy-paste step to reproduce.

**Comment by amereii** [comment]
@TeBoring , I do something like the following
```bash
git clone https://github.com/RooyeKhat-Media/iGap-API.git
cd iGap-API/app/assets/proto/
mkdir out
protoc --php_out=./out *.proto
```

**Comment by TeBoring** [comment]
I think you composer.json should be something like:
```
 "autoload": {
        "psr-4": {
            "":"the out directory of generated proto code"
        }
    },
```

**Comment by TeBoring** [comment]
Feel free to reopen if the problem still exist.

**Comment by amereii** [comment]
@TeBoring See https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/composer.json#L12-L18
Why you think it's wrong?

I tried the following and run `composer dumpautoload` and still having problem
```json
    "autoload": {
        "psr-4": {
            "iGap\\": "./src",
            "": "./src"
        }
    },
```

As **./src** contains GPBMetadata and Proto , https://github.com/RooyeKhat-Media/iGap-PHP-Client/tree/master/src

**Comment by TeBoring** [comment]
But the command you used to generate proto files direct to out directory

**Comment by amereii** [comment]
@TeBoring I just copy them to **./src**

**Comment by TeBoring** [comment]
You can not copy them. Code depends on the path.

**Comment by amereii** [comment]
@TeBoring Can you show me in these files
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/Proto/UserLogin.php
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/GPBMetadata/UserLogin.php

How is it depends on the path? It just use namespace ?

**Comment by TeBoring** [comment]
Why don’t you start with a simple proto and do some experiments with it.
See how to make it work.

**Comment by amereii** [comment]
@TeBoring Can you tell me why The following code works?

https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L25
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L32
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L41
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L55
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L74

You can see 
* ```new \Proto\UserRegister()```
* ```new ```
* ```new ```

**Comment by amereii** [comment]
@TeBoring Can you tell me why The following code works?

https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L25
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L33
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/register.php#L74

You can see 
* ```new \Proto\UserRegister()```
* ```new \Proto\UserVerify()```
* ```new \Proto\UserTwoStepVerificationVerifyPassword()```
Works well

But https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/examples/getRoomList.php#L44 ```new \Proto\UserLogin()``` cannot be constructed , Just  ```new \Proto\UserLogin()``` has **Segmentation fault** , If we have autoload issue why we have **Segmentation fault** it must be a fatal error

**Comment by TeBoring** [comment]
If you don't use protobuf c extension, is there any error?

**Comment by amereii** [comment]
@TeBoring If you mean I comment `extension=protobuf.so` and run script the error is

> PHP Fatal error:  Class 'Google\Protobuf\Internal\Message' not found in /home/client/src/Proto/UserLogin.php on line 14

**Comment by TeBoring** [comment]
You should first config your composer.json to work with protobuf’s php
implementation. Doesn’t it depend on protobuf in composer.json?

**Comment by amereii** [comment]
@TeBoring Can you explain how exactly? 
My composer.json https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/composer.json

**Comment by amereii** [comment]
I run `composer require google/protobuf` and comment `extension=protobuf.so` to disable c extension

When I use **Pure PHP package** I have the following error

> PHP Parse error:  syntax error, unexpected 'Global' (T_GLOBAL), expecting identifier (T_STRING) in /home/client/src/GPBMetadata/UserLogin.php on line 19

Here : https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/GPBMetadata/UserLogin.php#L19

**Comment by amereii** [comment]
I think as **Global** is reserved key in PHP we have problem

**Comment by amereii** [comment]
Google create a class named `global`
https://github.com/RooyeKhat-Media/iGap-PHP-Client/blob/master/src/GPBMetadata/Global.php#L7

**Comment by amereii** [comment]
You must handle reserved keys http://php.net/manual/en/reserved.keywords.php , I suggest prepend _

**Comment by TeBoring** [comment]
What's the protoc version you are using? In 3.6.0, the protoc has handled reserved name in proto filename

**Comment by amereii** [comment]
I opened issue 18 days ago and used libprotoc 3.5.1 , the 3.6.0 released 11 days ago , Let me try it

**Comment by amereii** [comment]
Solved in 3.6.0
